src/Framework/AspNetCoreAnalyzers/src/SourceGenerators/PublicTopLevelProgramGenerator.cs

```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNetCore.SourceGenerators;
+
+[Generator]
+public class PublicProgramSourceGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var internalGeneratedProgramClass = context.CompilationProvider.Select((compilation, cancellationToken) =>
+        {
+            var program = compilation.GetTypeByMetadataName("Program");
+            // If the program class is already public, we don't need to generate anything.
+            if (program is null || program.DeclaredAccessibility == Accessibility.Public)
```

Comment:
Might also want to confirm that it's not generic?

Reply:
Might also want to confirm that it contains a `Main` method?

Reply:
The `Main` method the compiler emits has a generated name by default. Perhaps we can rely on the semantics defined by the language to guarantee that a `Main` method exists?

Reply:
I'm not sure I understand what you mean by relying on the semantics of the language.  If I declare my own `class Program`, won't it be returned by `GetTypeByMetadataName`?  My questions were about confirming whether the type we found was actually the one generated for top-level statements.

Reply:
I'm referring to this bit in the spec:

>  It is a partial type, so a type named "Program" in source code must also be declared as partial.

It's not possible to declare your own type (interface, struct, etc.) named `Program` without getting a compiler error:

```
Console.WriteLine("foo");

internal struct Program { }
```

produces:

```
error CS0101: The namespace '<global namespace>' already contains a definition for 'Program'
```

It appears that when this is the case, `GetTypeByMetadataName("Program")` will return `null` so we already handle this in our first check.

Reply:
I had been assuming there might not be top-level statements, but we now check for that.